src/routing/comm.rs

```diff
@@ -122,295 +127,167 @@ impl Comm {
             );
         }
 
-        // Use `FuturesUnordered` to execute all the send tasks concurrently, but still on the same
-        // thread. Keep track of the sending progress using the `SendState` helper.
-        let mut state = SendState::new(recipients, delivery_group_size);
-        let mut tasks = FuturesUnordered::new();
-
-        loop {
-            // Start a batch of sends.
-            while let Some(addr) = state.next() {
-                let msg = msg.clone();
-                let task = async move {
-                    let result = self.send_once(&addr, msg).await;
-                    (addr, result)
-                };
-                tasks.push(task);
-            }
+        let delivery_group_size = delivery_group_size.min(recipients.len());
 
-            // Await until one of the started sends completes.
-            if let Some((addr, result)) = tasks.next().await {
-                // Notify `SendState` about the result of the send and potentially start the next
-                // send, appending to the ones still in progress (if any).
-                match result {
-                    Ok(_) => state.success(&addr),
-                    Err(err) => {
-                        trace!(
-                            "Sending message ({} bytes) to {} failed: {}",
-                            msg.len(),
-                            addr,
-                            err
-                        );
-                        state.failure(&addr);
+        // Run all the sends concurrently (using `FuturesUnordered`). If any of them fails, pick
+        // the next recipient and try to send to them. Proceed until the needed number of sends
+        // succeeds or if there are no more recipients to pick.
+        let send = |recipient, msg| async move { (self.send_to(recipient, msg).await, recipient) };
+
+        let mut tasks: FuturesUnordered<_> = recipients[0..delivery_group_size]
+            .iter()
+            .map(|recipient| send(recipient, msg.clone()))
+            .collect();
+
+        let mut next = delivery_group_size;
+        let mut successes = 0;
+        let mut failed_recipients = vec![];
+
+        while let Some((result, addr)) = tasks.next().await {
+            match result {
+                Ok(()) => successes += 1,
+                Err(qp2p::Error::Connection(qp2p::ConnectionError::LocallyClosed)) => {
+                    // The connection was closed by us which means we are terminating so let's cut
+                    // this short.
+                    return (Err(SendError), vec![]);
+                }
+                Err(_) => {
+                    failed_recipients.push(*addr);
+
+                    if next < recipients.len() {
```

Comment:
just pop(take) from recipients, so that using of `next` can be avoided?
also, when `next >= recipients.len()` , shall we break out of the loop?

Reply:
> just pop(take) from recipients

`recipients` is slice not `Vec`, so there is no `pop`. But even it there was, that would reverse the order of the recipients (but maybe that doesn't matter?). I could maybe do something with `slice::split_first` or so, but I don't think the code would be simpler.

> when next >= recipients.len() , shall we break out of the loop

We still need to wait for all the ongoing send task to finish.